csharp/src/Apache.Arrow/Ipc/ArrowStreamWriter.cs

```diff
@@ -399,6 +439,95 @@ private protected void WriteRecordBatchInternal(RecordBatch recordBatch)
             return Tuple.Create(recordBatchBuilder, fieldNodesVectorOffset);
         }
 
+
+        private protected void WriteDictionaries(RecordBatch recordBatch)
+        {
+            foreach (Field field in recordBatch.Schema.Fields.Values)
+            {
+                WriteDictionary(field);
+            }
+        }
+
+        private protected void WriteDictionary(Field field)
+        {
+            if (field.DataType.TypeId != ArrowTypeId.Dictionary)
+            {
+                if (field.DataType is NestedType nestedType)
+                {
+                    foreach (Field child in nestedType.Fields)
+                    {
+                        WriteDictionary(child);
+                    }
+                }
+                return;
+            }
+
+            (ArrowRecordBatchFlatBufferBuilder recordBatchBuilder, Offset<Flatbuf.DictionaryBatch> dictionaryBatchOffset) =
+                CreateDictionaryBatchOffset(field);
+
+            WriteMessage(Flatbuf.MessageHeader.DictionaryBatch,
+                dictionaryBatchOffset, recordBatchBuilder.TotalLength);
+
+            WriteBufferData(recordBatchBuilder.Buffers);
+        }
+
+        private protected async Task WriteDictionariesAsync(RecordBatch recordBatch, CancellationToken cancellationToken = default)
+        {
+            foreach (Field field in recordBatch.Schema.Fields.Values)
+            {
+                await WriteDictionaryAsync(field, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private protected async Task WriteDictionaryAsync(Field field, CancellationToken cancellationToken = default)
+        {
+            if (field.DataType.TypeId != ArrowTypeId.Dictionary)
+            {
+                if (field.DataType is NestedType nestedType)
+                {
+                    foreach (Field child in nestedType.Fields)
+                    {
+                        await WriteDictionaryAsync(child, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+                return;
+            }
+
+            (ArrowRecordBatchFlatBufferBuilder recordBatchBuilder, Offset<Flatbuf.DictionaryBatch> dictionaryBatchOffset) =
+                CreateDictionaryBatchOffset(field);
+
+            await WriteMessageAsync(Flatbuf.MessageHeader.DictionaryBatch,
+                dictionaryBatchOffset, recordBatchBuilder.TotalLength, cancellationToken).ConfigureAwait(false);
+
+            await WriteBufferDataAsync(recordBatchBuilder.Buffers, cancellationToken).ConfigureAwait(false);
+        }
+
+        private protected Tuple<ArrowRecordBatchFlatBufferBuilder, Offset<Flatbuf.DictionaryBatch>> CreateDictionaryBatchOffset(Field field)
```

Comment:
Can this just be `private`?

```suggestion
        private Tuple<ArrowRecordBatchFlatBufferBuilder, Offset<Flatbuf.DictionaryBatch>> CreateDictionaryBatchOffset(Field field)
```

Reply:
Fixed it.